File: src/main/webapp/www/js/dao/hino.dao.js
calvinApp.service('hinoDAO', ['database', '$q', function(database, $q){
        this.findUltimaAlteracaoHinos  = function(callback){
            database.db.transaction(function(tx) {
                tx.executeSql('SELECT max(ultima_atualizacao) as ultimaAtualizacao FROM hino', [], function(tx, rs) {
                    if (rs.rows && rs.rows.length){
                        callback(new Date(rs.rows.item(0).ultimaAtualizacao));
                    }else{
                        callback();
                    }
                }, function(tx, error) {
                    callback();
                });
            });
        };

        this.mergeHino = function(hino){
            database.db.transaction(function(tx) {
                tx.executeSql('delete from hino where id = ?', [hino.id]);

                tx.executeSql('insert into hino(id, numero, assunto, autor, nome, texto, filename, ultima_atualizacao) values(?,?,?,?,?,?,?,?)',
                [hino.id, hino.numero, hino.assunto, hino.autor, hino.nome, hino.texto, hino.filename, hino.ultimaAlteracao.getTime()]);
            });
        };

        this.findHinosByFiltro = function(filtro){
            var deferred = $q.defer();
            
            filtro = angular.extend({filtro:'',pagina:1,total:50}, filtro);

            database.db.transaction(function(tx) {
                tx.executeSql('SELECT id, numero, nome FROM hino where (numero || nome || texto) like ? order by numero, nome limit ?, ?', 
                            ['%' + filtro.filtro + '%', (filtro.pagina - 1) * filtro.total, filtro.total + 1], function(tx, rs) {
                    var hinos = {resultados:[],hasProxima:rs.rows.length > filtro.total};

                    for (var i=0;i<rs.rows.length && i<filtro.total;i++){
                        var item = rs.rows.item(i);
                        hinos.resultados.push({
                            id:item.id,
                            numero:item.numero,
                            nome:item.nome
                        });
                    }

                    deferred.resolve(hinos);
                }, function(tx, error) {
                    deferred.reject(error);
                });
            });

            return deferred.promise;
        };

        this.findHino = function(hino){
            var deferred = $q.defer();

            database.db.transaction(function(tx) {
                tx.executeSql('SELECT id, numero, nome, assunto, texto, autor, filename FROM hino where id = ?', [Number(hino)], function(tx, rs) {
                    if (rs.rows && rs.rows.length){
                        var item = rs.rows.item(0);
                        
                        deferred.resolve({
                          id:item.id,
                          numero:item.numero,
                          nome:item.nome,
                          assunto:item.assunto,
                          texto:item.texto,
                          autor:item.autor,
                          filename:item.filename
                        });
                    }else{
                        deferred.resolve(null);
                    }
                }, function(tx, error) {
                    deferred.reject(error);
                });
            });

            return deferred.promise;
        };

    }]);

File: lib/infra/dao/hino_dao.dart
part of pocket_church.infra;

class HinoDAO {
  Future<DateTime> findUltimaAlteracaoHinos() async {
    List<Map> rs = await pcDatabase.database.rawQuery(
        "SELECT max(ultima_atualizacao) as ultimaAtualizacao FROM hino");

    if (rs.isNotEmpty && rs[0]['ultimaAtualizacao'] != null) {
      return DateTime.fromMillisecondsSinceEpoch(rs[0]['ultimaAtualizacao']);
    }

    return null;
  }

  mergeHino(Hino hino) async {
    await pcDatabase.database.transaction((txn) async {
      await txn.execute(
        "delete from hino where id = ?",
        [hino.id],
      );

      await txn.execute(
        "insert into hino(id, numero, assunto, autor, nome, texto, filename, ultima_atualizacao) values(?,?,?,?,?,?,?,?)",
        [hino.id, hino.numero, hino.assunto, hino.autor, hino.nome, hino.texto, hino.filename, hino.ultimaAlteracao.millisecondsSinceEpoch],
      );

    });
  }

  Future<Pagina<Hino>> findHinosByFiltro({String filtro, int pagina = 1, int tamanhoPagina}) async {
    List<Map> rs = await pcDatabase.database.rawQuery(
      "SELECT id, numero, nome FROM hino where (numero || nome || texto) like ? order by numero, nome limit ?, ?",
      ['%' + (filtro ?? "") + '%', (pagina - 1) * tamanhoPagina, tamanhoPagina + 1],
    );

    return Pagina(
      pagina: pagina,
      hasProxima: rs.length > tamanhoPagina,
      resultados: rs.map((item) => Hino(
        id: item['id'],
        numero: item['numero'],
        nome: item['nome'],
      )).toList().sublist(0, math.min(rs.length, tamanhoPagina))
    );
  }

  Future<Hino> findHino(int id) async {
    List<Map> rs = await pcDatabase.database.rawQuery(
        "SELECT id, numero, nome, assunto, texto, autor, filename FROM hino where id = ?",
        [id]
    );

    if (rs.isNotEmpty) {
      Map item = rs[0];

      return Hino(
        id: item['id'],
        numero: item['numero'],
        nome: item['nome'],
        texto: item['texto'],
        assunto: item['assunto'],
        autor: item['autor'],
        filename: item['filename'],
      );
    }

    return null;
  }

}

HinoDAO hinoDAO = HinoDAO();
